front/post_upsert_hooks/hooks/document_tracker/actions/index.ts

```diff
@@ -0,0 +1,163 @@
+import {
+  Action,
+  cloneBaseConfig,
+  DustProdActionRegistry,
+} from "@app/lib/actions/registry";
+import { Authenticator, prodAPICredentialsForOwner } from "@app/lib/auth";
+import {
+  DustAPI,
+  DustAPIErrorResponse,
+  DustAppConfigType,
+} from "@app/lib/dust_api";
+import { Result } from "@app/lib/result";
+import {
+  ActionResponseBase,
+  DocTrackerLegacyActionMatchValue,
+  DocTrackerLegacyActionNoMatchValue,
+  DocTrackerLegacyActionSuccessResponse,
+  DocTrackerRetrievalActionSuccessResponse,
+  DocTrackerRetrievalActionValue,
+  isActionResponseBase,
+  isDocTrackerLegacyActionSuccessResponse,
+  isDocTrackerRetrievalActionSuccessResponse,
+} from "@app/post_upsert_hooks/hooks/document_tracker/actions/types";
+import { TRACKABLE_CONNECTOR_TYPES } from "@app/post_upsert_hooks/hooks/document_tracker/consts";
+
+export async function callLegacyDocTrackerAction(
+  workspaceId: string,
+  incomingDoc: string
+): Promise<
+  DocTrackerLegacyActionMatchValue | DocTrackerLegacyActionNoMatchValue
+> {
+  const action = DustProdActionRegistry["doc-tracker"];
+  const config = cloneBaseConfig(action.config);
+
+  config.SEMANTIC_SEARCH.data_sources = await getTrackableDataSources(
+    workspaceId
+  );
+
+  return callAction<DocTrackerLegacyActionSuccessResponse>({
+    workspaceId,
+    input: { incoming_document: incomingDoc },
+    action,
+    config,
+    responseChecker: isDocTrackerLegacyActionSuccessResponse,
+  });
+}
+
+export async function callDocTrackerRetrievalAction(
+  workspaceId: string,
+  inputText: string
+): Promise<DocTrackerRetrievalActionValue[]> {
+  const action = DustProdActionRegistry["doc-tracker-retrieval"];
+  const config = cloneBaseConfig(action.config);
+
+  config.SEMANTIC_SEARCH.data_sources = await getTrackableDataSources(
+    workspaceId
+  );
+
+  return callAction<DocTrackerRetrievalActionSuccessResponse>({
+    workspaceId,
+    input: { input_text: inputText },
+    action,
+    config,
+    responseChecker: isDocTrackerRetrievalActionSuccessResponse,
+  });
+}
+
+type ActionSuccessResponseBase = ActionResponseBase & {
+  results: { value: unknown }[][];
+};
+
+interface CallActionParams<R extends ActionSuccessResponseBase> {
+  workspaceId: string;
+  input: unknown;
+  action: Action;
+  config: DustAppConfigType;
+  responseChecker: (response: unknown) => response is R;
+}
+
+async function callAction<R extends ActionSuccessResponseBase>({
```

Comment:
Shouldn't we use a `Result` here?

What is `R["results"][0][0]["value"]` ?

Reply:
+1

Reply:
> What is R["results"][0][0]["value"]

It's the "value" part of the Dust API response

Reply:
Yes yes but it's super hard to parse as a reader of this code without a ton of context :)

Reply:
got it will try to make it clearer